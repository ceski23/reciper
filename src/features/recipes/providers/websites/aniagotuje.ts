import aniagotujeIcon from 'assets/images/providers/ania_gotuje.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode, isElementNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'

export const aniagotuje: RecipesProvider = {
	name: 'AniaGotuje.pl',
	icon: aniagotujeIcon,
	matcher: /https:\/\/aniagotuje\.pl\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe.image ? await getColorFromImage(partialRecipe.image) : undefined
		const name = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim()

		return {
			...partialRecipe,
			color,
			name,
			ingredients: parseIngredients(doc) ?? partialRecipe.ingredients,
			instructions: parseInstructions(doc) ?? partialRecipe.instructions,
			gallery: parseImages(doc),
		}
	},
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('[itemprop="recipeInstructions"] img'))
		.map(elem => elem.getAttribute('data-src'))
		.filter(isDefined)

const parseIngredients = (doc: Document) => {
	const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div')
	const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header')

	// Single recipe
	if (ingredientsHeaders?.length === 0) {
		const recipeIngredients = Array
			.from(root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text }))

		return recipeIngredients.length > 0 ? recipeIngredients : undefined
	}

	// Recipe with sub-recipes
	const recipeIngredients = Array.from(ingredientsHeaders ?? []).reduce<{
		ingredients: Recipe['ingredients']
		group?: string
	}>(({ ingredients, group }, element) => {
		const nextGroup = element.textContent?.trim() ?? group
		const nextIngredients = Array
			.from(element.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text, group: nextGroup }))

		return {
			ingredients: ingredients.concat(nextIngredients),
			group: nextGroup,
		}
	}, {
		ingredients: [],
		group: undefined,
	}).ingredients

	return recipeIngredients.length > 0 ? recipeIngredients : undefined
}

const parseInstructions = (doc: Document) => {
	const instructions: Recipe['instructions'] = []
	let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] div + :is(h3, h2)')
	let header: string | undefined

	while (currentElement !== null && currentElement !== undefined) {
		if (
			isElementNode(currentElement)
			&& ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
		) {
			header = currentElement.textContent?.trim()
		} else {
			const text = getTextFromNode(currentElement)

			if (text !== null && text.length > 0) {
				instructions.push({ text, group: header })
			}
		}

		currentElement = currentElement.nextSibling
	}

	return instructions.length > 0 ? instructions : undefined
}
