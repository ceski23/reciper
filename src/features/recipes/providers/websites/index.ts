export { aniagotuje } from './aniagotuje'
export { kwestiasmaku } from './kwestiasmaku'
export { wszystkiegoslodkiego } from './wszystkiegoslodkiego'
