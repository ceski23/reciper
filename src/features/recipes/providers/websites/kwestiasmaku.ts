import kwestiasmakuIcon from 'assets/images/providers/kwestia_smaku.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'
import { parseValidNumber } from 'lib/utils/numbers'

export const kwestiasmaku: RecipesProvider = {
	name: 'Kwestia Smaku',
	icon: kwestiasmakuIcon,
	matcher: /https:\/\/www\.kwestiasmaku\.com\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe?.image ? await getColorFromImage(partialRecipe.image).catch(() => undefined) : undefined
		const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim()

		const tags = Array
			.from(doc.querySelectorAll('#node-przepis-full-group-kategorie a'))
			.map(elem => elem.textContent?.toLowerCase())
			.filter(isDefined)

		const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji')
		const servings = parseValidNumber(servingsElement?.textContent)

		return {
			...partialRecipe,
			ingredients: parseIngredients(doc),
			instructions: parseInstructions(doc),
			color,
			description,
			tags,
			servings,
			gallery: parseImages(doc),
		}
	},
}

const parseIngredients = (doc: Document): Recipe['ingredients'] => {
	const ingredientsRoot = doc.querySelector('.field-name-field-skladniki')
	if (!ingredientsRoot) return []

	// Multiple ingredients groups
	if (ingredientsRoot.querySelector('.wyroznione')) {
		return Array.from(ingredientsRoot.children).reduce<{
			ingredients: Recipe['ingredients']
			group?: string
		}>(({ ingredients, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextIngredients = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				ingredients: ingredients.concat(nextIngredients),
				group: nextGroup,
			}
		}, {
			ingredients: [],
			group: undefined,
		}).ingredients
	}

	// Signle ingredients group
	const recipeIngredients = Array
		.from(ingredientsRoot.querySelectorAll('ul > li'))
		.map(elem => elem.textContent?.trim())
		.filter(isDefined)
		.map(text => ({ text }))

	return recipeIngredients
}

const parseInstructions = (doc: Document): Recipe['instructions'] => {
	const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie')
	if (!instructionsRoot) return []

	// Multiple instruction groups
	if (instructionsRoot.querySelector('.wyroznione')) {
		return Array.from(instructionsRoot.children).reduce<{
			instructions: Recipe['instructions']
			group?: string
		}>(({ instructions, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextInstructions = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				instructions: instructions.concat(nextInstructions),
				group: nextGroup,
			}
		}, {
			instructions: [],
			group: undefined,
		}).instructions
	}

	// Signle instructions group
	return Array
		.from(instructionsRoot.querySelectorAll('ul > li'))
		.map(elem => getTextFromNode(elem).trim())
		.filter(isDefined)
		.map(text => ({ text }))
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('.view-zdjecia img'))
		.map(elem => elem.getAttribute('src'))
		.filter(isDefined)
