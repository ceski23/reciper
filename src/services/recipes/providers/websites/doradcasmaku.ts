import icon from 'assets/providers/doradca_smaku.png';

import { Recipe } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';

export const DoradcaSmakuProvider: Provider = (() => {
  const name = 'Doradca smaku';
  const url = 'https://www.doradcasmaku.pl/';

  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    const recipeImageElem = doc.querySelector('.recipe_item .recipe_gallery .img_wrapper img');
    const image = recipeImageElem?.getAttribute('data-src') ?? undefined;

    const partial: Partial<Recipe> = {
      image,
    };

    return Object.assign(data, partial);
  };

  return {
    name, url, scrapper, icon,
  };
})();
