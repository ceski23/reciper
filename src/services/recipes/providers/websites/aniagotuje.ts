import icon from 'assets/providers/ania_gotuje.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { isElementNode, nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const AniaGotujeProvider: Provider = (() => {
  const name = 'AniaGotuje.pl';
  const url = 'https://aniagotuje.pl/';

  const parseIngredients = (doc: Document) => {
    const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div');
    const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header');
    const ingredients: RecipeIngredient[] = [];

    // Single recipe
    if (ingredientsHeaders?.length === 0) {
      const groupItemsElement = root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text }));

      ingredients.push(...groupItems);

      return ingredients;
    }

    // Recipe with sub-recipes
    ingredientsHeaders?.forEach((ingHeader) => {
      const groupTitle = ingHeader.textContent?.trim();

      const groupItemsElement = ingHeader.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text, group: groupTitle }));

      ingredients.push(...groupItems);
    });

    return ingredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructions: RecipeInstruction[] = [];
    let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] .recipe-ing-list + h2');
    let header: string | undefined;

    while (currentElement !== null && currentElement !== undefined) {
      if (
        isElementNode(currentElement)
        && ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
      ) {
        header = currentElement.textContent?.trim();
      } else {
        const text = getTextFromNode(currentElement);

        if (text !== null && text.length > 0) {
          instructions.push({ text, group: header });
        }
      }

      currentElement = currentElement.nextSibling;
    }

    return instructions;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await microdataScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const recipeName = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim();

    const partial: Partial<Recipe> = {
      color,
      name: recipeName,
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
