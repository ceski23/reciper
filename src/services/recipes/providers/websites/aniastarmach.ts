import icon from 'assets/providers/ania_starmach.png';

import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { colorExtractor } from 'services/recipes/providers/utils';

export const AniaStarmachProvider: Provider = (() => {
  const name = 'Ania Starmach';
  const url = 'https://aniastarmach.pl/';

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const tagsElements = doc.querySelectorAll('.tags-wrapper a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(Boolean) as string[];

    return {
      ...data,
      color,
      tags,
    };
  };

  return {
    name, url, scrapper, icon,
  };
})();
