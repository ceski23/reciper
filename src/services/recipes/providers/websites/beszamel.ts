/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/beszamel.svg';

import { Recipe, RecipeIngredient } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const BeszamelProvider: Provider = (() => {
  const name = 'Beszamel';
  const url = 'https://beszamel.se.pl/';

  const parseIngredients = (doc: Document) => {
    const elements = doc.querySelectorAll('.ingredients__items li');
    let group: string | undefined;
    const ingredients: RecipeIngredient[] = [];

    for (const element of elements) {
      const text = getTextFromNode(element);

      if (text.endsWith(':')) {
        group = text;
      } else {
        ingredients.push(removeEmpty({ group, text }));
      }
    }

    return ingredients;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const tagsElements = doc.querySelectorAll('.breadcrumb .breadcrumb__item');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase().trim())
      .filter(nonNullable);

    const partial: Partial<Recipe> = {
      color,
      tags,
      ingredients: parseIngredients(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
