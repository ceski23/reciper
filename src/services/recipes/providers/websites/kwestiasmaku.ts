/* eslint-disable @typescript-eslint/no-loop-func */
/* eslint-disable @typescript-eslint/brace-style */
/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/kwestia_smaku.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:kwestiasmaku');

export const KwestiaSmakuProvider: Provider = (() => {
  const name = 'Kwestia Smaku';
  const url = 'https://www.kwestiasmaku.com/';

  const parseIngredients = (doc: Document) => {
    const ingredientsRoot = doc.querySelector('.field-name-field-skladniki');
    if (!ingredientsRoot) return [];

    // Multiple ingredients groups
    if (ingredientsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const ingredients: RecipeIngredient[] = [];

      for (const element of ingredientsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeIngredients = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => elem.textContent?.trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          ingredients.push(...recipeIngredients);
        }
      }

      return ingredients;
    }

    // Signle ingredients group
    const ingredientsElements = ingredientsRoot.querySelectorAll('ul > li');
    const recipeIngredients = Array
      .from(ingredientsElements)
      .map((elem) => elem.textContent?.trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeIngredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie');
    if (!instructionsRoot) return [];

    // Multiple instruction groups
    if (instructionsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const instructions: RecipeInstruction[] = [];

      for (const element of instructionsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeInstructions = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => getTextFromNode(elem).trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          instructions.push(...recipeInstructions);
        }
      }

      return instructions;
    }

    // Signle instructions group
    const instructionsElements = instructionsRoot.querySelectorAll('ul > li');
    const recipeInstructions = Array
      .from(instructionsElements)
      .map((elem) => getTextFromNode(elem).trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeInstructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('.view-zdjecia img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const microdata = await microdataScrapper(doc);
    const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim() || undefined;

    const tagsElements = doc.querySelectorAll('#node-przepis-full-group-kategorie a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji');
    const servings = servingsElement?.textContent
      ? Number.parseInt(servingsElement.textContent, 10)
      : undefined;

    const color = microdata.image ? await getColorFromImage(microdata.image) : undefined;

    const partial: Partial<Recipe> = {
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      color,
      description,
      tags,
      servings,
      gallery: parseImages(doc),
    };

    return Object.assign(microdata, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
