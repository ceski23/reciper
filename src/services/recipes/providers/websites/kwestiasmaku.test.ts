import { expect } from 'vitest';

import { ciastoKruche, lemoniada } from 'services/recipes/providers/fixtures/kwestiasmaku';
import { KwestiaSmakuProvider } from 'services/recipes/providers/websites/kwestiasmaku';

const scrapeRecipe = async (data: string) => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(data, 'text/html');
  const partialRecipe = await KwestiaSmakuProvider.scrapper(doc);
  partialRecipe.id = '1';

  return partialRecipe;
};

describe('scrapper', () => {
  it('should scrape valid recipe', async () => {
    const recipe = await scrapeRecipe(lemoniada);
    expect(recipe).toMatchSnapshot();
  });

  it('should scrape valid recipe with sub-recipes', async () => {
    const recipe = await scrapeRecipe(ciastoKruche);
    expect(recipe).toMatchSnapshot();
  });
});
