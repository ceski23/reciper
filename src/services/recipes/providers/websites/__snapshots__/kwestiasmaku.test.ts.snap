// Vitest Snapshot v1

exports[`scrapper > should scrape valid recipe 1`] = `
{
  "color": "#7f7f7f",
  "description": "To sposób na pyszny napój dla całej rodziny. Świetnie sprawdzi się na przyjęciach, ładnie wygląda i każdy zawsze sięga po dokładkę :-)
                      Pomarańcze i cytryny kroję na półplasterki, zasypuję cukrem i odstawiam do lodówki. Dodaję też nasiona kardamonu (wcześniej podprażone aby uwolniły swój aromat) oraz liście mięty. Taką mieszankę trzymam w lodówce do czasu podania lemoniady. Wówczas do cytrusów dolewam zimną wodę, świeżo wyciśnięty sok z cytryny, pomarańczy i limonki. Gotową lemoniadę podaję z kostkami lodu.
                      Lato na szczęście ciągle trwa, zapowiada się słoneczna pogoda na następne dni, warto skorzystać z przepisu i zrobić dobrą lemoniadę; świetnie sprawdzi się na stole np. podczas weekendowego grilla ;-)",
  "id": "1",
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/lemoniada_00.jpg",
  "ingredients": [
    {
      "text": "3 cytryny",
    },
    {
      "text": "1 limonka (lub 1 dodatkowa cytryna)",
    },
    {
      "text": "2 pomarańcze",
    },
    {
      "text": "7 łyżek cukru",
    },
    {
      "text": "1 litr wody niegazowanej",
    },
    {
      "text": "1 gałązka mięty",
    },
    {
      "text": "opcjonalnie: 4 nasiona kardamonu",
    },
  ],
  "instructions": [
    {
      "text": "Cytrusy dokładnie wyszorować, następnie przekroić na połówki. Odkroić po jednej ćwiartce z 1 pomarańczy, 1 cytryny i limonki i pokroić je na plasterki. Włożyć do dzbanka, zasypać cukrem.",
    },
    {
      "text": "Z reszty owoców wycisnąć sok, np. za pomocą wyciskarki, pozbywając się pestek i nadmiaru miąższu. Sok przelać do dzbanka z owocami.",
    },
    {
      "text": "Dodać wodę i gałązkę mięty. Dokładnie wymieszać. W razie potrzeby można dodać więcej wody i cukru (zależy od tego jak soczyste były cytryny).",
    },
  ],
  "name": "Lemoniada",
  "rating": 4.84,
  "servings": 4,
  "tags": [
    "napoje",
    "cytryny",
    "pomarańcze",
    "przyjęcia i imprezy",
    "dla dzieci",
    "sylwester",
    "na upały",
    "kardamon",
  ],
}
`;

exports[`scrapper > should scrape valid recipe with sub-recipes 1`] = `
{
  "color": "#7f7f7f",
  "description": "Ciasto kruche z truskawkami, serem i kruszonką. Tradycyjne ciasto w nowoczesnej odsłonie. Często pytacie o modyfikację tego ulubionego przepisu - czy da się w prosty sposób zamienić rabarbar na truskawki? Jak je przygotować aby ciasto nie namokło od wilgotnych owoców? W tym przepisie rozwiąże Wasze wszelkie wątpliwości.
                     Mocną stroną tego wypieku - oprócz pysznej warstwy serowej i kokosowej kruszonki - jest idealne, kruche i delikatne ciasto z dodatkiem mąki ryżowej lub ziemniaczanej. Mąka ryżowa - kiedyś trudna do kupienia - jest teraz w stałej ofercie większości supermarketów w świetnej cenie. Spróbujcie jej w cieście kruchym a nie będziecie używać już innej :-) Oczywiście możecie użyć też łatwiejszej do dostania mąki ziemniaczanej.",
  "id": "1",
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/ciasto_kruche_z_truskawkami_i_serem_00.jpg",
  "ingredients": [
    {
      "group": "Ciasto:",
      "text": "150 g mąki pszennej",
    },
    {
      "group": "Ciasto:",
      "text": "100 g mąki ryżowej lub ziemniaczanej",
    },
    {
      "group": "Ciasto:",
      "text": "50 g cukru",
    },
    {
      "group": "Ciasto:",
      "text": "150 g masła (zimnego)",
    },
    {
      "group": "Ciasto:",
      "text": "1 jajko",
    },
    {
      "group": "Masa serowa:",
      "text": "500 g zmielonego twarogu",
    },
    {
      "group": "Masa serowa:",
      "text": "1 budyń waniliowy lub śmietankowy lub 40 g mąki ziemniaczanej",
    },
    {
      "group": "Masa serowa:",
      "text": "100 g cukru",
    },
    {
      "group": "Masa serowa:",
      "text": "2 jajka",
    },
    {
      "group": "Masa serowa:",
      "text": "100 ml śmietanki kremówki (zimnej)",
    },
    {
      "group": "Kruszonka:",
      "text": "150 g mąki pszennej",
    },
    {
      "group": "Kruszonka:",
      "text": "50 g mąki ryżowej lub ziemniaczanej",
    },
    {
      "group": "Kruszonka:",
      "text": "50 g wiórków kokosowych",
    },
    {
      "group": "Kruszonka:",
      "text": "50 g cukru",
    },
    {
      "group": "Kruszonka:",
      "text": "125 g masła (zimnego)",
    },
    {
      "group": "Owoce:",
      "text": "500 g truskawek",
    },
    {
      "group": "Owoce:",
      "text": "słoiczek (około 250 g) konfitury truskawkowej gotowej lub domowej",
    },
    {
      "group": "Owoce:",
      "text": "3/4 szklanki soku lub syropu lub intensywnego kompotu truskawkowego",
    },
    {
      "group": "Owoce:",
      "text": "1/3 szklanki mąki ziemniaczanej lub ryżowej (50 g)",
    },
  ],
  "instructions": [
    {
      "group": "Ciasto:",
      "text": "Formę o wymiarach około 20 x 30 cm posmarować masłem i wyłożyć papierem do pieczenia, tak aby wystawał poza brzegi formy.",
    },
    {
      "group": "Ciasto:",
      "text": "Do miski przesiać obydwie mąki, dodać cukier, pokrojone na kawałeczki zimne masło i rozcierać palcami (lub mieszadłem miksera) aż powstanie drobna kruszonka.",
    },
    {
      "group": "Ciasto:",
      "text": "Dodać jajko i szybko zagnieść ciasto łącząc składniki w jednolitą kulę. Rozpłaszczyć ją, zawinąć w folię i włożyć do lodówki na około 60 minut (w międzyczasie przygotować kruszonkę, truskawki i masę serową).",
    },
    {
      "group": "Ciasto:",
      "text": "Ciastem wyłożyć dno i około 1,5 cm boków formy. Spód podziurkować widelcem i wstawić do lodówki na około 30 minut. Piekarnik nagrzać do 160 stopni C. Piec przez 30 minut na złoty kolor. Wyjąć ciasto, ale nie wyłączać piekarnika.",
    },
    {
      "group": "Kruszonka:",
      "text": "Do miski przesiać obydwie mąki, dodać wiórki kokosowe, cukier oraz pokrojone na kawałeczki zimne masło. Rozcierać palcami masło z mąką aż powstanie drobna kruszonka. Jeśli jest bardzo sypka, drobna, można ugnieść trochę kruszonkę w dłoni aby miejscami złączyła się w większe kawałki ciasta. Wstawić do lodówki.",
    },
    {
      "group": "Owoce:",
      "text": "Truskawki opłukać, osuszyć, oderwać szypułki, pokroić na kawałki. Włożyć do garnka razem ze zmiksowaną blenderem konfiturą. Podgrzewać przez około 5 minut co chwilę mieszając aż owoce zmiękną. Sok truskawkowy wymieszać z mąką ziemniaczaną i wlać do truskawek, wymieszać, chwilę podgrzać i jak tylko zgęstnieje odstawić z ognia i wyłożyć do czystej miski i ostudzić.",
    },
    {
      "group": "Masa serowa:",
      "text": "Do miski włożyć ser, dodać budyń (proszek) i cukier, wymieszać łyżką lub krótko zmiksować. Dodać jajka i wymieszać na jednolitą masę lub zmiksować. Śmietankę ubić na sztywną pianę i dodać do masy serowej, delikatnie wymieszać łyżką.",
    },
    {
      "group": "Masa serowa:",
      "text": "Na podpieczony spód wyłożyć masę serową i wstawić z powrotem do piekarnika. Piec przez 15 - 20 minut aż masa serowa się zetnie. Wyjąć ciasto z piekarnika i zwiększyć temp. w piekarniku do 180 stopni C. Na warstwę serową wyłożyć truskawki i posypać kruszonką. Wstawić do piekarnika na 35 minut.",
    },
  ],
  "name": "Ciasto kruche z truskawkami i serem",
  "rating": 4.45,
  "tags": [
    "truskawki",
    "ciasta kruche",
    "ciasta z truskawkami",
  ],
}
`;
